src/topp/GNPSExport.cpp

```diff
@@ -79,297 +147,368 @@ using namespace std;
-            if (cosine_sim >= cos_sim)
+          if (cos_sim >= cos_sim_threshold)
+          {
+            for (auto spec_iter = test_spec.begin(); spec_iter < test_spec.end(); spec_iter++)
             {
-              for (Size spectrum_index = 0; spectrum_index < spectrum.size(); ++spectrum_index)
-              {
-                // exp.addSpectrum(spectrum);
-                // exp.addSpectrum(spectrum);
-                auto curr_spectrum = spectrum[spectrum_index];
-                if (ms2_block[curr_spectrum.getMZ()] < curr_spectrum.getIntensity())
-                {
-                  ms2_block[curr_spectrum.getMZ()] = curr_spectrum.getIntensity();
-                }
-              }
+              mz_int_pairs.emplace_back(spec_iter->getMZ(), spec_iter->getIntensity());
             }
           }
+        }
+        sort(mz_int_pairs.begin(), mz_int_pairs.end());
 
+        map<double,int> ms2_block;
+        generateMSMSSpectrumBins(mz_int_pairs, bin_width, ms2_block);
 
-          feature_stream << "BEGIN IONS" << endl;
-
-          feature_stream << "FEATURE_ID=" << feature_count++ << endl;
-          feature_stream << "CONSENSUSID=e_" << feature.getUniqueId() << endl;
+        // write output
+        writeMS2BlockToFile(output_file, ms2_block, output_type, (cons_i+1), feature.getUniqueId(),
+                            charge, feature.getMZ(), best_speci, best_spec.getRT());
 
-          feature_stream << "MSLEVEL=2" << endl;
-          feature_stream << "CHARGE=" << std::to_string(charge == 0 ? 1 : charge) << "+" << endl;
-          feature_stream << "PEPMASS=" << peptides[0].first.second.getMZ() << endl;
-          feature_stream << "FILE_INDEX=" << peptides[0].second.second << endl;
-          feature_stream << "RTINSECOND=" << peptides[0].first.second.getRT() << endl;
+        // delete allocated resources
+        mz_int_pairs.clear();         
+      
 
-          for (auto ms2_iter = ms2_block.rbegin(); ms2_iter != ms2_block.rend(); ++ms2_iter)
-          {
-            feature_stream << ms2_iter->first << "\t" << (int) ms2_iter->second << endl;
-          }
-          feature_stream << "END IONS" << endl << endl;
-        }
-
-        // output feature information to general outputStream
-        output_stream << feature_stream.str() << endl;
+        element_maps.clear();
+        pepts.clear();
       }
-    }
+      
+      progress_logger.setProgress(cons_i);
+    } // end of for-loop across features
     progress_logger.endProgress();
 
-    //-------------------------------------------------------------
-    // writing output
-    //-------------------------------------------------------------
-    ofstream output_file(out);
-    progress_logger.startProgress(0, 1, "writing mgf file");
-    output_file << output_stream.str();
-    progress_logger.endProgress();
+    // delete / dealloc mem resources
     output_file.close();
 
+    mzml_file_paths.clear();
```

Comment:
no need to clear at the end of a function.
The destructor will take care of it automatically.